cmd/controller/cmd/init_certs.go

```diff
@@ -100,23 +102,51 @@ func runCertsCmd(ctx context.Context, propellerCfg *config.Config, cfg *webhook.
 
 func createWebhookSecret(ctx context.Context, namespace string, cfg *webhook.Config, certs webhookCerts, secretsClient v1.SecretInterface) error {
 	isImmutable := true
-	_, err := secretsClient.Create(ctx, &corev1.Secret{
+	secretData := map[string][]byte{
+		CaCertKey:            certs.CaPEM.Bytes(),
+		ServerCertKey:        certs.ServerPEM.Bytes(),
+		ServerCertPrivateKey: certs.PrivateKeyPEM.Bytes(),
+	}
+
+	secret := &corev1.Secret{
 		ObjectMeta: v12.ObjectMeta{
 			Name:      cfg.SecretName,
 			Namespace: namespace,
 		},
-		Type: corev1.SecretTypeOpaque,
-		Data: map[string][]byte{
-			CaCertKey:            certs.CaPEM.Bytes(),
-			ServerCertKey:        certs.ServerPEM.Bytes(),
-			ServerCertPrivateKey: certs.PrivateKeyPEM.Bytes(),
-		},
+		Type:      corev1.SecretTypeOpaque,
+		Data:      secretData,
 		Immutable: &isImmutable,
-	}, v12.CreateOptions{})
+	}
+
+	_, err := secretsClient.Create(ctx, secret, v12.CreateOptions{})
 
 	if errors.IsAlreadyExists(err) {
-		// TODO: Maybe get the secret and validate it has all the required keys?
-		logger.Infof(ctx, "A secret already exists with the same name. Ignoring creating secret.")
+		logger.Infof(ctx, "A secret already exists with the same name. Validating.")
+		s, err := secretsClient.Get(ctx, cfg.SecretName, v12.GetOptions{})
+		if err != nil {
+			return err
+		}
+
+		// If any key is missing from the secret, replace all of them. There is no value in only filling in the missing
+		// keys since they all need to work in tandem.
+		requiresUpdate := false
+		for key := range secretData {
+			if _, exists := s.Data[key]; !exists {
+				requiresUpdate = true
+				break
+			}
+		}
+
+		if requiresUpdate {
```

Comment:
we should probably log that we are updating things, as it may lead to confusion?

Reply:
maybe we can add an annotation, saying system updated?